core/src/main/java/org/apache/iceberg/hadoop/HadoopMetricsContext.java

```diff
@@ -114,12 +111,18 @@ public void increment(Integer amount) {
     };
   }
 
-  private void writeObject(ObjectOutputStream out) throws IOException {
-    out.defaultWriteObject();
-  }
+  private FileSystem.Statistics statistics() {
+    FileSystem.Statistics localStatisticsRef = statistics;
```

Comment:
why do we need a `localStatisticsRef`?

If we already have `private transient FileSystem.Statistics statistics`, I think we can just check against that and see:

```java
    if (statistics == null) {
      synchronized (this) {
        if (statistics == null) {
          this.statistics = FileSystem.getStatistics(scheme, null);
        }
      }
    }

return statistics;
```

I think this is also used in places like https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/SerializableTable.java#L117-L132

Reply:
agree with you in this context it seems to be un-necessary, have made the changes for the same.

generally we use this format in singleton pattern where there is a volatile obj holding the reference of the singleton. (https://www.javacodemonk.com/threadsafe-singleton-design-pattern-java-806ad7e6).


  | Using localRef, we are reducing the access of volatile  variable to just one for positive usecase. If we do not use localRef,  then we would have to access volatile variable twice - once for checking  null and then at method return time. Accessing volatile memory is quite  an expensive affair because it involves reaching out to main memory.
-- | --

Reply:
>  then we would have to access volatile variable twice - once for checking null and then at method return time

I agree, but the second access is only done for one time for initialization, so I think we can take the cost in this case

Reply:
came across this blog : https://www.journaldev.com/171/thread-safety-in-java-singleton-classes
This states that having this kind of localRef's can increase perf of this method as much as 25% 

"The local variable result seems unnecessary. But, it’s there to improve the performance of our code. In cases where the instance is already initialized (most of the time), the volatile field is only accessed once (due to “return result;” instead of “return instance;”). This can improve the method’s overall performance by as much as 25 percent."

I think we will access it always 2 times first to check if it;s initialized and the last time to actually return it.